File: pytorch/modules/WeightNormalizedLinear.py
import math
import torch
from torch.nn.parameter import Parameter
import torch.nn.functional as F
from torch.nn.modules.module import Module

class WeightNormalizedLinear(Module):

    def __init__(self, in_features, out_features, scale=True, bias=True, init_factor=1):
        super(WeightNormalizedLinear, self).__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.weight = Parameter(torch.Tensor(out_features, in_features))
        if bias:
            self.bias = Parameter(torch.zeros(1, out_features))
        else:
            self.register_parameter('bias', None)
        if scale:
            self.scale = Parameter(torch.ones(1, out_features))
        else:
            self.register_parameter('scale', None)
        self.reset_parameters(init_factor)

    def reset_parameters(self, factor):
        stdv = 1. * factor / math.sqrt(self.weight.size(1))
        self.weight.data.uniform_(-stdv, stdv)
        if self.bias is not None:
            self.bias.data.uniform_(-stdv, stdv)

    def weight_norm(self):
        return self.weight.pow(2).sum(1).add(1e-6).sqrt()

    def norm_scale_bias(self, input):
        output = input.div(self.weight_norm().transpose(0, 1).expand_as(input))
        if self.scale is not None:
            output = output.mul(self.scale.expand_as(input))
        if self.bias is not None:
            output = output.add(self.bias.expand_as(input))
        return output

    def forward(self, input):
        return self.norm_scale_bias(F.linear(input, self.weight))

    def __repr__(self):
        return self.__class__.__name__ + ' (' \
            + str(self.in_features) + ' -> ' \
            + str(self.out_features) + ')'

File: torch/modules/WeightNormalizedLinear.lua
local WNL, parent = torch.class('nn.WeightNormalizedLinear', 'nn.Module')

function WNL:__init(nInput, nOutput, hasScale, hasBias, initWeightFactor)
	parent.__init(self)
	if hasScale == nil then
		self.hasScale = true
	else
		self.hasScale = hasScale
	end
	if hasBias == nil then
		self.hasBias = true
	else
		self.hasBias = hasBias
	end
	self.nInput = nInput
	self.nOutput = nOutput
	self.weight = torch.Tensor(nInput, nOutput)
	stdv = initWeightFactor / math.sqrt(nInput)
	self.weight:apply(function()
		return torch.uniform(-stdv, stdv)
	end)
	self.gradWeight = torch.zeros(nInput, nOutput)
	if self.hasScale then
		self.scale = torch.ones(nOutput)
		self.gradScale = torch.zeros(nOutput)
	end
	if self.hasBias then
		self.bias = torch.zeros(nOutput)
		self.gradBias = torch.zeros(nOutput)
	end
	self.eps = 1e-8
	self:updateStats()
end

function WNL:updateStats()
	self.weightNorm = torch.sqrt(torch.add(torch.pow(self.weight, 2):sum(1), self.eps))
end

function WNL:updateOutput(input)
	self.tmp = torch.cdiv(torch.mm(input, self.weight), self.weightNorm:expand(input:size(1), self.nOutput))
	self.output = self.tmp
	if self.hasScale then
		self.output:cmul(self.scale:reshape(1, self.nOutput):expand(input:size(1), self.nOutput))
	end
	if self.hasBias then
		self.output:add(self.bias:reshape(1, self.nOutput):expand(input:size(1), self.nOutput))
	end
	return self.output
end

function WNL:updateGradInput(input, gradOutput)
	self.gradTmp = gradOutput
	if self.hasScale then
		self.gradTmp:cmul(self.scale:reshape(1, self.nOutput):expand(input:size(1), self.nOutput))
	end
	self.gradInput = torch.mm(torch.cdiv(self.gradTmp, self.weightNorm:expand(input:size(1), self.nOutput)), self.weight:t())
	return self.gradInput
end

function WNL:accGradParameters(input, gradOutput, scale)
	self.gradWeight:add(
		scale,
		torch.cmul(
			torch.bmm(
				input:reshape(input:size(1), self.nInput, 1),
				torch.pow(self.weightNorm, -1):reshape(1, 1, self.nOutput):expand(input:size(1), 1, self.nOutput)
			)
			-
			torch.cmul(
				self.weight:reshape(1, self.nInput, self.nOutput):expand(input:size(1), self.nInput, self.nOutput),
				torch.cdiv(
					self.tmp,
					torch.pow(self.weightNorm, 2):reshape(1, self.nOutput):expand(input:size(1), self.nOutput)
				):reshape(input:size(1), 1, self.nOutput):expand(input:size(1), self.nInput, self.nOutput)
			),
			self.gradTmp:reshape(input:size(1), 1, self.nOutput):expand(input:size(1), self.nInput, self.nOutput)
		):sum(1)[1]
	)
	if self.hasScale then
		self.gradScale:add(scale, torch.cmul(self.tmp, gradOutput):sum(1)[1])
	end
	if self.hasBias then
		self.gradBias:add(scale, gradOutput:sum(1)[1])
	end
end

function WNL:zeroGradParameters()
	self.gradWeight:zero()
	if self.hasScale then
		self.gradScale:zero()
	end
	if self.hasBias then
		self.gradBias:zero()
	end
end

function WNL:updateParameters(learningRate)
	self.weight:add(-learningRate, self.gradWeight)
	if self.hasScale then
		self.scale:add(-learningRate, self.gradScale)
	end
	if self.hasBias then
		self.bias:add(-learningRate, self.gradBias)
	end
end

function WNL:type(__type)
	self.weight = self.weight:type(__type)
	self.gradWeight = self.gradWeight:type(__type)
	if self.hasScale then
		self.scale = self.scale:type(__type)
		self.gradScale = self.gradScale:type(__type)
	end
	if self.hasBias then
		self.bias = self.bias:type(__type)
		self.gradBias = self.gradBias:type(__type)
	end
	self.weightNorm = self.weightNorm:type(__type)
end

function WNL:parameters()
	local param = {self.weight}
	local grad = {self.gradWeight}
	if self.hasScale then
		table.insert(param, self.scale)
		table.insert(grad, self.gradScale)
	end
	if self.hasBias then
		table.insert(param, self.bias)
		table.insert(grad, self.gradBias)
	end
	return param, grad
end

function WNL:clearState()
	self.output = nil
	self.gradInput = nil
	self.tmp = nil
	self.gradTmp = nil
end